pkg/ipamd/ipamd.go

```diff
@@ -579,66 +609,70 @@ func (c *IPAMContext) tryUnassignIPsFromAll() {
 			}
 
 			// Delete IPs from datastore
-			var deletedIPs []string
+			var deletedIPsOrPrefixes []string
 			for _, toDelete := range ips {
 				// Don't force the delete, since a freeable IP might have been assigned to a pod
 				// before we get around to deleting it.
-				err := c.dataStore.DelIPv4AddressFromStore(eniID, toDelete, false /* force */)
+				err := c.dataStore.CleanupDataStore(eniID, toDelete, false /* force */, c.enableIpv4PrefixDelegation)
+
 				if err != nil {
 					log.Warnf("Failed to delete IP %s on ENI %s from datastore: %s", toDelete, eniID, err)
 					ipamdErrInc("decreaseIPPool")
 					continue
 				} else {
-					deletedIPs = append(deletedIPs, toDelete)
+					deletedIPsOrPrefixes = append(deletedIPsOrPrefixes, toDelete)
 				}
-			}
 
-			// Deallocate IPs from the instance if they aren't used by pods.
-			if err := c.awsClient.DeallocIPAddresses(eniID, deletedIPs); err != nil {
-				log.Warnf("Failed to decrease IP pool by removing IPs %v from ENI %s: %s", deletedIPs, eniID, err)
-			} else {
-				log.Debugf("Successfully decreased IP pool by removing IPs %v from ENI %s", deletedIPs, eniID)
+				// Deallocate IPs from the instance if they aren't used by pods.
+				if err = c.awsClient.DeallocIPAddresses(eniID, deletedIPsOrPrefixes, false); err != nil {
```

Comment:
I might be confused by the diff, but I think this needs to be moved out a level.  If I'm reading it correctly, I _think_ you're deallocating `deletedIPsOrPrefixes` inside the same loop as where you're building that list.

Reply:
oh good catch, merge messed it up :(